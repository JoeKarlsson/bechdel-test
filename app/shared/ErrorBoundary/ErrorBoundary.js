import React from 'react';
import PropTypes from 'prop-types';

class ErrorBoundary extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			hasError: false,
			error: null,
			errorInfo: null,
		};
	}

	componentDidCatch(error, info) {
		// Display fallback UI
		this.setState({
			hasError: true,
			error,
			errorInfo: info,
		});
		// You can also log the error to an error reporting service
		console.error(error, info);
	}

	render() {
		if (this.state.hasError) {
			// You can render any custom fallback UI
			return (
				<div className="Error">
					<h1>Something went wrong.</h1>
					<details style={{ whiteSpace: 'pre-wrap' }}>
						{this.state.error && this.state.error.toString()}
						<br />
						{this.state.errorInfo.componentStack}
					</details>
				</div>
			);
		}
		return this.props.children;
	}
}

ErrorBoundary.propTypes = {
	children: PropTypes.any, // eslint-disable-line react/forbid-prop-types
};

ErrorBoundary.defaultProps = {
	children: '',
};

export default ErrorBoundary;
