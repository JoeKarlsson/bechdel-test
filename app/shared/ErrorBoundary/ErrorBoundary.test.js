import React from 'react';
import {
	shallow,
	configure,
} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ErrorBoundary from './ErrorBoundary';

configure({ adapter: new Adapter() });

describe('Error Boundary', () => {
	let wrapper;
	let inst;

	beforeEach(() => {
		wrapper = shallow(
			<ErrorBoundary>
				When you buy a lottery ticket, you are investing in the dreams of the winner.
			</ErrorBoundary>,
		);
		inst = wrapper.instance();
	});

	describe('rendering', () => {
		describe('initial state', () => {
			it('should match the snapshot', () => {
				const component = renderer.create(
					<ErrorBoundary>
						When you buy a lottery ticket, you are investing in the dreams of the winner.
					</ErrorBoundary>,
				);
				const tree = component.toJSON();
				expect(tree).toMatchSnapshot();
			});

			it('is rendered correctly', () => {
				expect(wrapper).toHaveLength(1);
			});

			it('should render to static HTML', () => {
				expect(wrapper.text()).toContain('When you buy a lottery ticket, you are investing in the dreams of the winner.');
			});

			it('should have correct inital state', () => {
				const initialState = inst.state;
				const expectedIntialState = {
					hasError: false,
				};
				expect(initialState).toMatchObject(expectedIntialState);
			});

			it('should not have any inital props', () => {
				const initialProps = inst.props;
				const expectedProps = {
					children: 'When you buy a lottery ticket, you are investing in the dreams of the winner.',
				};
				expect(initialProps).toMatchObject(expectedProps);
			});
		});
	});

	describe('callbacks', () => {

		class BuggyComponent extends React.Component {
			componentDidMount() {
				throw new Error('I crashed!');
			}
			render() {
				return <h1>Buggy Component</h1>;
			}
		}

		it('should only render error message when `hasError` is true', () => {

			wrapper = shallow(
				<ErrorBoundary>
					<BuggyComponent />
				</ErrorBoundary>,
			);

			wrapper
				.instance()
				.componentDidCatch({ toString: () => 'error' }, { componentStack: { toString: () => 'info' } });

			wrapper.update();
			expect(wrapper.text()).toBe('Something went wrong.error');
		});

	});
});
