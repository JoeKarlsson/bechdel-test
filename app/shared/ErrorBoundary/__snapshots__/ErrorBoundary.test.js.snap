// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Boundary rendering initial state should match the snapshot 1`] = `"When you buy a lottery ticket, you are investing in the dreams of the winner."`;
